File: src/main/kotlin/io/github/dylmeadows/eontimer/service/factory/timer/DelayTimer.kt
package io.github.dylmeadows.eontimer.service.factory.timer

import io.github.dylmeadows.eontimer.model.TimerState
import io.github.dylmeadows.eontimer.model.settings.TimerSettingsModel
import io.github.dylmeadows.eontimer.model.timer.TimerConstants
import io.github.dylmeadows.eontimer.service.CalibrationService
import io.github.dylmeadows.eontimer.util.milliseconds
import io.github.dylmeadows.eontimer.util.reactor.FluxFactory
import io.github.dylmeadows.eontimer.util.toMinimumLength
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import java.time.Duration
import kotlin.math.absoluteValue

@Service
class DelayTimer @Autowired constructor(
    private val secondTimer: SecondTimer,
    private val timerSettings: TimerSettingsModel,
    private val calibrationService: CalibrationService) {

    fun createStages(targetSecond: Long, targetDelay: Long, calibration: Long): List<Duration> {
        return listOf(
            stage1(targetSecond, targetDelay, calibration),
            stage2(targetDelay, calibration))
    }

    fun createTimer(targetSecond: Long, targetDelay: Long, calibration: Long): Flux<TimerState> {
        return FluxFactory.fixedTimer(timerSettings.refreshInterval.milliseconds,
            createStages(targetSecond, targetDelay, calibration))
    }

    fun calibrate(targetDelay: Long, delayHit: Long): Long {
        val delta = calibrationService.toMillis(delayHit) - calibrationService.toMillis(targetDelay)
        return when {
            delta.absoluteValue <= TimerConstants.CLOSE_THRESHOLD ->
                (TimerConstants.CLOSE_UPDATE_FACTOR * delta).toLong()
            else ->
                (TimerConstants.UPDATE_FACTOR * delta).toLong()
        }
    }

    private fun stage1(targetSecond: Long, targetDelay: Long, calibration: Long): Duration {
        return (secondTimer.createStages(calibration, targetSecond)[0].toMillis()
            - calibrationService.toMillis(targetDelay)).toMinimumLength()
            .milliseconds
    }

    private fun stage2(targetDelay: Long, calibration: Long): Duration {
        return (calibrationService.toMillis(targetDelay) - calibration)
            .milliseconds
    }
}

File: src/services/timer/DelayTimer.cpp
//
// Created by Dylan Meadows on 2020-03-12.
//

#include "DelayTimer.h"
#include <util/Functions.h>

namespace service::timer {
    const int CLOSE_THRESHOLD = 167;
    const double UPDATE_FACTOR = 1.0;
    const double CLOSE_UPDATE_FACTOR = 0.75;

    DelayTimer::DelayTimer(const CalibrationService *calibrationService,
                           const SecondTimer *secondTimer)
        : calibrationService(calibrationService),
          secondTimer(secondTimer) {
    }

    const std::shared_ptr<std::vector<int>>
    DelayTimer::createStages(const int targetSecond, const int targetDelay, const int calibration) const {
        std::shared_ptr<std::vector<int>> stages = std::make_shared<std::vector<int>>(2);
        (*stages)[0] = createStage1(targetSecond, targetDelay, calibration);
        (*stages)[1] = createStage2(targetDelay, calibration);
        return stages;
    }

    const int DelayTimer::createStage1(const int targetSecond, const int targetDelay, const int calibration) const {
        return util::functions::toMinimumLength(
            secondTimer->createStage1(targetSecond, calibration) -
            calibrationService->toMilliseconds(targetDelay)
        );
    }

    const int DelayTimer::createStage2(const int targetDelay, const int calibration) const {
        return calibrationService->toMilliseconds(targetDelay) - calibration;
    }

    int DelayTimer::calibrate(const int targetDelay, const int delayHit) const {
        const int delta = calibrationService->toMilliseconds(delayHit) -
                          calibrationService->toMilliseconds(targetDelay);
        if (std::abs(delta) <= CLOSE_THRESHOLD) {
            return static_cast<int>(CLOSE_UPDATE_FACTOR * delta);
        } else {
            return static_cast<int>(UPDATE_FACTOR * delta);
        }
    }
}
